### ISSUE TITLE
Implement Automatic Language Detection When Typing

### ISSUE THREAD

**Issue by glufke** [futo-org/android-keyboard#478]
### Use Cases

It's great. But there's one feature missing: it should automatically change the language according to what I type. Example: I have 2 languages installed. Sometimes I write in English, and sometimes not... The Google keyboard detects what language I'm typing, probably by the first or 2 words I type. And then it sticks with that language. That's the only thing it should be improved in order to ditch forever Google keyboard... Thanks!!!!

**Comment by Madderose** [comment]
Same, I wish there was a dual input, or at least a language selector based on application used to start with. I switch all the time between two languages and it's such a pain...

**Comment by Chinoman10** [comment]
Dual input would be a blessing!

On Swiftkey I'd always have either English-Portuguese or English-French, as I often use English words in the middle of other-language conversations, and it works really well.

**Comment by ExecuteOrd66** [comment]
I can't stress the importance of this feature!

**Comment by PatatePwel** [comment]
Living in a bilingual country, I unfortunately had to uninstall the FUTO keyboard due to the lack of this feature. It is indeed essential.

**Comment by JamarDoyle** [comment]
Yes please, its the 1 thing i need then this would work very well for me and alot of other people i know

**Comment by Chinoman10** [comment]
> Living in a bilingual country, I unfortunately had to uninstall the FUTO keyboard due to the lack of this feature. It is indeed essential.

Every day I'm stressing out because FUTO's keyboard isn't bilingual... I'm constantly switching between Portuguese with family and friends and English for business, and I feel like I'm getting white hairs by the day...
It's a shame having to switch to Swiftkey after paying for the FUTO keyboard to help support this great FOSS initiative... but in it's current state I unfortunately cannot use it anymore and can't recommend it either.

I'll keep a close eye and wait for new developments to hopefully come back in the future.

**Comment by haldi4803** [comment]
I think that Is a really important feature.

But because it might take some times until it gets implemented. How are we gonna do it ourselves?

According to docs:
https://gitlab.futo.org/keyboard/keyboard-wiki/-/wikis/Keyboard-LM-docs

It should be possible to create our own dictionary which is bilingual by default or not? Might totally kill all correct predictions.....
Anyone tried before?

Or.... We train our own Transformer model.
https://gitlab.futo.org/keyboard/keyboard-wiki/-/wikis/Keyboard-LM-docs#can-you-support-xyz-language-for-transformer-lm

I'm not sure if the automatically "Fine-tune" option in settings works for multiple languages... Probably not...

So do it yourself:
https://gitlab.futo.org/keyboard/keyboard-wiki/-/wikis/Keyboard-LM-docs#model-creation

Either train our own custom Model. Or try to create a LORA to the existing model for another language... Not sure how well that might work.
Maybe not at all...

**Comment by haldi4803** [comment]
Original Topic was #38

**Comment by MaLoLHD** [comment]
> It should be possible to create our own dictionary which is bilingual by default or not? Might totally kill all correct predictions..... Anyone tried before?
I spent some time trying to make a transformer following the code snippets provided in the docs, but I couldn't get it working.
The docs mention:
> TODO: Training notebooks
If notebooks were released to make training models trivial, I think that would fix the problem.

**Comment by enricocirene** [comment]
It'd be such a dream for the keyboard to automatically switch between languages, it is the only missing piece for me as well. An amazing app without it, but it'd be perfect with.

**Comment by Nordlicht-13** [comment]
It would help, when you just have two languages, to change between them by longpressing the spacebar, without the extra dialog for choosing the language.

**Comment by dean-futo** [comment]
Very similar to #461 - merged that issue into this one.

**Comment by UnconsciousBlaringHuman** [comment]
> It would help, when you just have two languages, to change between them by longpressing the spacebar, without the extra dialog for choosing the language.

FWIW, you can replace the Emoji button with a Language switch key, which allows you to switch without long pressing or showing you that dialog. Just go to `Keyboard & Typing` > `Edit Actions` then place `Language switch key` in the `Action Key` section and the `Emojis` key somewhere else.

**Comment by dean-futo** [comment]
Very similar to #683 - merging this request into that one.

**Comment by Chinoman10** [comment]
@dean-futo FWIW, I believe issue #478 and #461 are different.
- #461 is about supporting multiple languages **simultaneously**, like Gboard or Microsoft's Swiftkey keyboards.
  - I believe this is the ideal solution, effectively merging multiple (_maybe just 2 for now_) languages into "one", so you can "dual-wield" two languages at the same time.
- #478 (_this one_) is about automatically detecting what kind of language is being written, and automatically switch to that language; a perhaps easier implementation than 461, but definitely not the same UX.

**Comment by codethief** [comment]
@Chinoman10 [Great minds…](https://github.com/futo-org/android-keyboard/issues/683#issuecomment-2453567069) :) Anyway, @dean-futo has re-opened the present ticket. Meanwhile, I believe #461 (closed) is the same as #683 (still open), so I believe all should be good now? (Maybe with the exception of #708 still needing to be closed.)

**Comment by dean-futo** [comment]
@Chinoman10 thanks for elaborating. Yes, they are a different UX related to multi-lingual support and ultimately we will need to experiment to see which methods work best. Some languages are easier to implement than others and these multi-lingual support requests will be a slow rollout.

**Comment by dean-futo** [comment]
@codethief yes, they should all be consolidated enough at this point to get at the core separate types of requests, all related to multi-lingual support.

**Comment by maddler** [comment]
This is the single feature preventing me to use Futo Keyboard full time. Finger crissed it'll make ti the app soon

**Comment by Wesselinator** [comment]
Just adding on here, this is the biggest pain point for FUTO and why I haven't suggested it to my friends yet.

---

More then half[*](https://www.reddit.com/r/languagelearning/comments/yimsc0/comment/iujs2vm) of the word is multi-lingual, but only about [20%](https://www.census.gov/library/stories/2022/12/languages-we-speak-in-united-states.html) of the USA speaks more then one language. 

I consider the above important to point out to everyone here: FUTO is a USA company where they expect this feature to only really be useful to 1 out of every 5 of their local users. 

But this is also a chicken-&-egg problem, I know I can't convince my friends to switch off of SwiftKey or GBoard because they consider multi-lingual typing as a "default" feature and will not be able live without it.
I assume then that most (if not all) multilingual people will consider FUTO not worth it and will use something else.  

Why implement this if multilingual people are not using your keyboard? Why use this keyboard if this feature is not there?

---

# My Experience

Talking as a native bilingual to other native bilinguals involves using the words your brain grabs first which is very often **not** the language you were speaking. **It actually take a non trivial amount of effort to stay in one language.**
The linguistic term for this is [Code Switching](https://en.wikipedia.org/wiki/Code-switching)

Some people have been suggesting an easier language switch button, but that is not good enough in my opinion. 
Being spoiled on GBoard's multi lingual typing, I'll try to explain my experience

## GBoard's Multi Lingual Keyboard

_They allowed 2 languages but I think there was a way to add up to 4._

The way GBoard does it is very very good in my opinion. You pick 2 languages, a primary one and a secondary one:
*  as you are typing it assumes Language A and suggests auto completes for it in the bar
*  only if the word is completely impossible to exist in A, would it suggest a word from B.
* If you misspelled a word in language B it would still auto correct it for you.
* this auto correcting was biased towards your primary language for words spelled similarly
* depending on the last word you wrote it would bias the next suggestions in the direction of the word.
> e.g. " AAA AAAAAA AAAAAA BBBB BBB" would result in a bias towards B with at least one suggestion in B
> note: as A was still your "Primary" language it would only suggest the single most likely word in B, very rarely would it suggest a second word for B.

I had `A-B` and `B-A` setup and depending on the language I wanted more suggestions in, I would switch appropriately.

---

# Technical speculation

Just want to add on here some technical speculation:

There are multiple ways they could have implemented this:

## Creative mixing of Language models

They could have figured out a creative way to mix the outputs of two language models (+ dictionaries) without them clashing.

This is the most pragmatic way to do it as the models of the languages on their own get better they will make multi lingual better.

```mermaid
flowchart TD
    A(Language A) -- ^% --> C[Outputs]
    B(Language B) -- v% --> C
```

This means you can mix and match any language and the system will be able to deal with it.

## Multi Lingual layer

They could have another model sitting onto of the the 2 as in:
```mermaid
flowchart TD
    A(Language A) --> C{A-B}
    B(Language B) --> C
    C --> D[Outputs]
```

Which gets trained on the petabytes of typing data google has + training data from people using multi-lingual typing.

This solution suffers from the fact you need a intermediate model for every combination of languages.
> You might even need one for the inverse!  
> I know I type differently in A-B vs. B-A so it might not be as simple as doing a `* -1` on the outputs.

This would be a "future proof" solution for google as it would be constantly retrained from the typing data of how users use it.

## Dedicated Models

Maybe they had enough compute and data to train entire language combinations.  
This would be _insane_ to do for a feature you consider [less important](https://support.google.com/gboard/?hl=en#topic=) then [Morse Code Typing](https://support.google.com/gboard/?hl=en#topic=9023899)

## Straight up just mixing the dictionaries and using a generic model 

It might just be as simple as that with heavier weights given to the "primary" language.

**Comment by djlalle** [comment]
I can not switch to FUTO keyboard for the same reason.
Usually I write in three languages all the time and I can not press always the space bar to switch language :)
unfortunately I have to uninstall it

**Comment by brog-io** [comment]
Same, FUTO would be perfect if this feature exists. Is there any update on this? I'll use heliboard right now

**Comment by kittenvr** [comment]
Wow 5 months and still not implemented

**Comment by kittenvr** [comment]
> > Wow 5 months and still not implemented
> 
> This takes a lot of time to get right. Apple and Google probably took longer than that to figure it out. Also, FUTO has to achieve this with a technique that is less privacy invasive.

Heliboard got it right and ia opensource

**Comment by jerji** [comment]
> > > Wow 5 months and still not implemented
> Heliboard got it right and ia opensource
This is already marked as high priority, if you want to look at the other
tool and make a pull request for review, you are welcome to do so :)

**Comment by JdotCarver** [comment]
Yes, please do not complain on speed of implementation if you aren't willing to provide any help.
I also think this feature is very necessary, as english is a secondary language to most non-anglophone coutries.

**Comment by alensiljak** [comment]
It seems that a lot of people here are limiting their thinking to bilingual function. As a regular user of at least three languages, I hope this gets implemented as a multi-lingual feature and not just bilingual.
Cheers for a great keyboard app.

**Comment by abb128** [comment]
As multilingual typing has been implemented in nightly (see issue #683) I will be closing this issue as I believe it should effectively address this use case and there isn't much need for a language detection separate feature

**Comment by codethief** [comment]
@abb128 Would you mind re-opening this ticket for now? Multilingual typing has completely different constraints and will never be as accurate as typing in a single (auto-detected) language. See also the previous discussion in https://github.com/futo-org/android-keyboard/issues/478#issuecomment-2454032376 .

**Comment by biohazardxxx** [comment]
Would it be possible to reopen this?

Multilingual typing helps with autocorrect and predictions, but it doesn't cover the cases where the layout itself needs to change (different alphabets, or layouts with different key sets like QWERTZ vs QWERTY). For those, manual switching is still the only option.

Automatic detection based on what's being typed would solve this. Even a conservative version would help - e.g. only switch after a few clearly-identified words, and never mid-word, so it doesn't fight the user.

There's also a simpler variant that would cover a lot of the same ground: remembering the last used language per app. Most people write in one language in their work chat and another in their family chat, so a per-app default would remove some of the manual switching without needing any detection logic at all.

Happy to help test if this gets picked up.

**Comment by abb128** [comment]
@biohazardxxx For the first part, I don't think it makes much sense at all for a keyboard to switch layouts automatically in the middle of typing. If you're multilingual English/Russian for example, you can't even begin to type a Russian word in QWERTY in the first place. This is the case for any language pair that use different alphabets. Even for the same alphabet, I doubt that it'd be useful as you'd have to write German words with QWERTY and expect the Y/Z to suddenly get swapped at some point and you may as well just stick with one layout, plus multilingual typing already adds the necessary accents or diacritics of other languages into long pressing keys. This is unless all of the recent input events get reinterpreted under a different layout and past nonsense words get rewritten, which gets into being a really complex rabbit hole that I would rather not get into.

For the second part, there's an issue open for that #709
